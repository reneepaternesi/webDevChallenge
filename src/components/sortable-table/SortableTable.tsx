import React, { useState, useMemo } from 'react'
import { useHistory } from 'react-router-dom'

import {
  TableWrapper,
  TableHead,
  TableHeadRow,
  TableHeadElement,
  TableHeadBtn,
  TableBody,
  TableBodyRow,
  TableBodyCell,
} from './styles'

const useSortableData = (items: any, config?: any) => {
  const [sortConfig, setSortConfig] = useState(config)

  const sortedItems = useMemo(() => {
    let sortableItems = [...items]
    if (sortConfig) {
      sortableItems.sort((a, b) => {
        if (a[sortConfig?.key] < b[sortConfig.key]) {
          return sortConfig.direction === 'ascending' ? -1 : 1
        }
        if (a[sortConfig.key] > b[sortConfig.key]) {
          return sortConfig.direction === 'ascending' ? 1 : -1
        }
        return 0
      })
    }
    return sortableItems
  }, [items, sortConfig])

  const requestSort = (key: any) => {
    let direction = 'ascending'
    if (sortConfig && sortConfig.key === key && sortConfig.direction === 'ascending') {
      direction = 'descending'
    }
    setSortConfig({ key, direction })
  }

  return { items: sortedItems, requestSort, sortConfig }
}

interface Props {
  results: Array<{
    name: string
    rotation_period: string
    orbital_period: string
    diameter: string
    population: string
  }>
}

const SortableTable: React.FC<Props> = ({ results }) => {
  const { items, requestSort, sortConfig } = useSortableData(results)
  const history = useHistory()

  const getClassNamesFor = (name: string) => {
    if (!sortConfig) {
      return
    }
    return sortConfig.key === name ? sortConfig.direction : undefined
  }

  const navigateTo = (item: any) => history.push({
    pathname: '/planet',
    state: { planet: item }
  })

  return (
    <TableWrapper>
      <TableHead>
        <TableHeadRow>
          <TableHeadElement>
            <TableHeadBtn type="button" onClick={() => requestSort('name')} className={getClassNamesFor('name')}>
              Name
            </TableHeadBtn>
          </TableHeadElement>
          <TableHeadElement>
            <TableHeadBtn
              type="button"
              onClick={() => requestSort('rotation_period')}
              className={getClassNamesFor('rotation_period')}
            >
              Rotation Period
            </TableHeadBtn>
          </TableHeadElement>
          <TableHeadElement>
            <TableHeadBtn
              type="button"
              onClick={() => requestSort('orbital_period')}
              className={getClassNamesFor('orbital_period')}
            >
              Orbital Period
            </TableHeadBtn>
          </TableHeadElement>
          <TableHeadElement>
            <TableHeadBtn
              type="button"
              onClick={() => requestSort('diameter')}
              className={getClassNamesFor('diameter')}
            >
              Diameter
            </TableHeadBtn>
          </TableHeadElement>
          <TableHeadElement>
            <TableHeadBtn
              type="button"
              onClick={() => requestSort('population')}
              className={getClassNamesFor('population')}
            >
              Population
            </TableHeadBtn>
          </TableHeadElement>
        </TableHeadRow>
      </TableHead>
      <TableBody>
        {items.map((item, index) => (
          <TableBodyRow key={index} onClick={() => navigateTo(item)}>
            <TableBodyCell>{item.name}</TableBodyCell>
            <TableBodyCell>{item.rotation_period}</TableBodyCell>
            <TableBodyCell>${item.orbital_period}</TableBodyCell>
            <TableBodyCell>{item.diameter}</TableBodyCell>
            <TableBodyCell>{item.population}</TableBodyCell>
          </TableBodyRow>
        ))}
      </TableBody>
    </TableWrapper>
  )
}

export default SortableTable
