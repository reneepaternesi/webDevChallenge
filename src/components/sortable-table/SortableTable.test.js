import React from 'react'
import { shallow } from 'enzyme'
import SortableTable from './SortableTable'

describe('SortableTable Test', () => {
  it('should render SortableTable correctly', () => {
    const props = {
      results: [
          {
            "name": "Geonosis", 
            "rotation_period": "30", 
            "orbital_period": "256", 
            "diameter": "11370", 
            "population": "100000000000"
          },
          {
            "name": "Utapau", 
            "rotation_period": "27", 
            "orbital_period": "351", 
            "diameter": "12900", 
            "population": "95000000", 
          }
      ]
    }

    const component = shallow(<SortableTable {...props} />)
    expect(component).toMatchSnapshot()
  })
})
