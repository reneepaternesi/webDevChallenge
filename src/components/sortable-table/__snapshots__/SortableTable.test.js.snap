// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortableTable Test should render SortableTable correctly 1`] = `
<styled.table>
  <styled.thead>
    <styled.tr>
      <styled.th>
        <styled.button
          onClick={[Function]}
          type="button"
        >
          Name
        </styled.button>
      </styled.th>
      <styled.th>
        <styled.button
          onClick={[Function]}
          type="button"
        >
          Rotation Period
        </styled.button>
      </styled.th>
      <styled.th>
        <styled.button
          onClick={[Function]}
          type="button"
        >
          Orbital Period
        </styled.button>
      </styled.th>
      <styled.th>
        <styled.button
          onClick={[Function]}
          type="button"
        >
          Diameter
        </styled.button>
      </styled.th>
      <styled.th>
        <styled.button
          onClick={[Function]}
          type="button"
        >
          Population
        </styled.button>
      </styled.th>
    </styled.tr>
  </styled.thead>
  <styled.tbody>
    <styled.tr
      key="0"
      onClick={[Function]}
    >
      <styled.td>
        Geonosis
      </styled.td>
      <styled.td>
        30
      </styled.td>
      <styled.td>
        $
        256
      </styled.td>
      <styled.td>
        11370
      </styled.td>
      <styled.td>
        100000000000
      </styled.td>
    </styled.tr>
    <styled.tr
      key="1"
      onClick={[Function]}
    >
      <styled.td>
        Utapau
      </styled.td>
      <styled.td>
        27
      </styled.td>
      <styled.td>
        $
        351
      </styled.td>
      <styled.td>
        12900
      </styled.td>
      <styled.td>
        95000000
      </styled.td>
    </styled.tr>
  </styled.tbody>
</styled.table>
`;
